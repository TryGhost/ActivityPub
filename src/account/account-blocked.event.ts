export class AccountBlockedEvent {
    constructor(
        private readonly accountId: number,
        private readonly blockerId: number,
    ) {}

    getAccountId(): number {
        return this.accountId;
    }

    getBlockerId(): number {
        return this.blockerId;
    }

    static getName(): string {
        return 'account.blocked';
    }
}
