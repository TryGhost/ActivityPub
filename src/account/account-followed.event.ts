import type { Account } from './account.entity';

export class AccountFollowedEvent {
    constructor(
        private readonly account: Account,
        private readonly follower: Account,
    ) {}

    getAccount(): Account {
        return this.account;
    }

    getFollower(): Account {
        return this.follower;
    }

    static getName(): string {
        return 'account.followed';
    }
}
