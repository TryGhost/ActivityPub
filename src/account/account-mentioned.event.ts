import type { Post } from 'post/post.entity';

export class AccountMentionedEvent {
    constructor(
        private readonly post: Post,
        private readonly accountId: number,
    ) {}

    getPost(): Post {
        return this.post;
    }

    getAccountId(): number {
        return this.accountId;
    }

    static getName(): string {
        return 'account.mentioned';
    }
}
