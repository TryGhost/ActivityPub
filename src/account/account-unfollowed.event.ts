export class AccountUnfollowedEvent {
    constructor(
        private readonly accountId: number,
        private readonly unfollowerId: number,
    ) {}

    getAccountId(): number {
        return this.accountId;
    }

    getUnfollowerId(): number {
        return this.unfollowerId;
    }

    getName(): string {
        return 'account.unfollowed';
    }

    static getName(): string {
        return 'account.unfollowed';
    }
}
