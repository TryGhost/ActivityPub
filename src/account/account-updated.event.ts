import type { Account } from './account.entity';

export class AccountUpdatedEvent {
    static getName(): string {
        return 'account.updated';
    }

    constructor(private readonly account: Account) {}

    getAccount(): Account {
        return this.account;
    }
}
