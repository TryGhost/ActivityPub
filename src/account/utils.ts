import { type Actor, PropertyValue } from '@fedify/fedify';
import type { ExternalAccountData } from './types';

interface PublicKey {
    id: string;
    owner: string;
    publicKeyPem: string;
}

/**
 * Map a Fedify Actor to an external account data object
 *
 * @param actor Actor
 */
export async function mapActorToExternalAccountData(
    actor: Actor,
): Promise<ExternalAccountData> {
    const customFields: Record<string, string> = {};

    for await (const attachment of actor.getAttachments()) {
        if (!(attachment instanceof PropertyValue)) {
            continue;
        }

        const name = attachment.name?.toString() || '';
        const value = attachment.value?.toString() || '';

        if (name && value) {
            customFields[name] = value;
        }
    }

    let apPublicKey: PublicKey | null = null;

    const publicKey = await actor.getPublicKey();

    if (publicKey) {
        const jsonLd = (await publicKey.toJsonLd({
            format: 'compact',
        })) as Partial<PublicKey>;

        if (typeof jsonLd === 'object' && jsonLd !== null) {
            apPublicKey = {
                id: jsonLd.id ?? '',
                owner: jsonLd.owner ?? '',
                publicKeyPem: jsonLd.publicKeyPem ?? '',
            };
        }
    }

    return {
        username: actor.preferredUsername?.toString() ?? '',
        name: actor.name?.toString() ?? null,
        bio: actor.summary?.toString() ?? null,
        avatar_url: (await actor.getIcon())?.url?.toString() ?? null,
        banner_image_url: (await actor.getImage())?.url?.toString() ?? null,
        url: actor.url?.toString() ?? null,
        custom_fields:
            Object.keys(customFields).length > 0 ? customFields : null,
        ap_id: actor.id?.href ?? '',
        ap_inbox_url: actor.inboxId?.href ?? '',
        ap_shared_inbox_url: actor.endpoints?.sharedInbox?.href ?? null,
        ap_outbox_url: actor.outboxId?.href ?? '',
        ap_following_url: actor.followingId?.href ?? '',
        ap_followers_url: actor.followersId?.href ?? '',
        ap_liked_url: actor.likedId?.href ?? '',
        ap_public_key: apPublicKey ? JSON.stringify(apPublicKey) : '',
    };
}

/**
 * Compute the handle for an account from the provided host and username
 *
 * @param host Host of the site the account belongs to
 * @param username Username of the account
 */
export function getAccountHandle(host?: string, username?: string) {
    return `@${username || 'unknown'}@${host || 'unknown'}`;
}
