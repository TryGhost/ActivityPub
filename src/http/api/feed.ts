import type { AccountService } from '../../account/account.service';
import { getAccountHandle } from '../../account/utils';
import type { AppContext } from '../../app';
import type { FeedService, FeedType } from '../../feed/feed.service';
import type { Post } from './types';

/**
 * Default number of posts to return in a feed
 */
const DEFAULT_FEED_POSTS_LIMIT = 20;

/**
 * Maximum number of posts that can be returned in a feed
 */
const MAX_FEED_POSTS_LIMIT = 100;

/**
 * Create a handler to handle a request for a user's feed
 *
 * @param feedService Feed service instance
 * @param accountService Account service instance
 * @param feedType Type of feed
 */
export function createGetFeedHandler(
    feedService: FeedService,
    accountService: AccountService,
    feedType: FeedType,
) {
    /**
     * Handle a request for a user's feed
     *
     * @param ctx App context
     */
    return async function handleGetFeed(ctx: AppContext) {
        const queryCursor = ctx.req.query('next');
        const cursor = queryCursor ? decodeURIComponent(queryCursor) : null;

        const queryLimit = ctx.req.query('limit');
        const limit = queryLimit
            ? Number(queryLimit)
            : DEFAULT_FEED_POSTS_LIMIT;

        if (limit > MAX_FEED_POSTS_LIMIT) {
            return new Response(null, {
                status: 400,
            });
        }

        const account = await accountService.getDefaultAccountForSite(
            ctx.get('site'),
        );

        const { results, nextCursor } = await feedService.getFeedData({
            accountId: account.id,
            feedType,
            limit,
            cursor,
        });

        const posts: Post[] = results.map((result) => {
            return {
                id: result.post_ap_id,
                type: result.post_type,
                title: result.post_title ?? '',
                excerpt: result.post_excerpt ?? '',
                content: result.post_content ?? '',
                url: result.post_url,
                featureImageUrl: result.post_image_url ?? null,
                publishedAt: result.post_published_at,
                likeCount: result.post_like_count,
                likedByMe: result.post_liked_by_user === 1,
                replyCount: result.post_reply_count,
                readingTimeMinutes: result.post_reading_time_minutes,
                attachments: [], // TODO
                author: {
                    id: result.author_id.toString(),
                    handle: getAccountHandle(
                        result.author_url
                            ? new URL(result.author_url).host
                            : '',
                        result.author_username,
                    ),
                    name: result.author_name ?? '',
                    url: result.author_url ?? '',
                    avatarUrl: result.author_avatar_url ?? '',
                },
                repostCount: result.post_repost_count,
                repostedByMe: result.post_reposted_by_user === 1,
                repostedBy: result.reposter_id
                    ? {
                          id: result.reposter_id.toString(),
                          handle: getAccountHandle(
                              result.reposter_url
                                  ? new URL(result.reposter_url).host
                                  : '',
                              result.reposter_username,
                          ),
                          name: result.reposter_name ?? '',
                          url: result.reposter_url ?? '',
                          avatarUrl: result.reposter_avatar_url ?? '',
                      }
                    : null,
            };
        });

        return new Response(
            JSON.stringify({
                posts,
                next: nextCursor,
            }),
            {
                status: 200,
            },
        );
    };
}
