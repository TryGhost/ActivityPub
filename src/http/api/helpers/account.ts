import { type Context, isActor } from '@fedify/fedify';
import type { Account } from 'account/account.entity';
import type { AccountService } from 'account/account.service';
import type { Site } from 'account/types';
import { getAccountHandle } from 'account/utils';
import type { ContextData } from 'app';
import {
    getAttachments,
    getFollowerCount,
    getFollowingCount,
    getHandle,
    isFollowedByDefaultSiteAccount,
} from 'helpers/activitypub/actor';
import { sanitizeHtml } from 'helpers/html';
import { lookupObject } from 'lookup-helpers';
import type { AccountDTO } from '../types';

/**
 * Converts an Account to an AccountDTO
 * @param account - The account to convert
 *
 * @returns Promise resolving to AccountDTO
 */
export async function getAccountDtoFromAccount(
    account: Account,
    defaultAccount: Account,
    accountService: AccountService,
): Promise<AccountDTO> {
    const accountDto: AccountDTO = {
        id: account.id?.toString() || '',
        name: account.name || '',
        handle: getAccountHandle(new URL(account.apId).host, account.username),
        bio: sanitizeHtml(account.bio || ''),
        url: account.url.toString() || '',
        avatarUrl: account.avatarUrl?.toString() || '',
        /**
         * At the moment we don't support banner images for Ghost accounts
         */
        bannerImageUrl: account.bannerImageUrl?.toString() || '',
        /**
         * At the moment we don't support custom fields for Ghost accounts
         */
        customFields: {},
        attachment: [],
        postCount: await accountService.getPostCount(account.id),
        likedCount: await accountService.getLikedCount(account.id),
        followingCount: await accountService.getFollowingAccountsCount(
            account.id,
        ),
        followerCount: await accountService.getFollowerAccountsCount(
            account.id,
        ),
        followedByMe: await accountService.checkIfAccountIsFollowing(
            defaultAccount.id,
            account.id,
        ),
        followsMe: false,
    };
    return accountDto;
}

/**
 * Retrieves an AccountDTO by looking up an actor using their handle
 * @param handle - The handle to look up (e.g., "@username@domain")
 *
 * @returns Promise resolving to AccountDTO
 */
export async function getAccountDTOByHandle(
    handle: string,
    apCtx: Context<ContextData>,
    site: Site,
    accountService: AccountService,
): Promise<AccountDTO> {
    if (!handle) {
        throw new Error('Handle is null');
    }

    // Lookup actor by handle
    const actorObj = await lookupObject(apCtx, handle);

    if (!isActor(actorObj)) {
        throw new Error('Actor not found');
    }

    const actor: any = await actorObj.toJsonLd();

    const [followerCount, followingCount, isFollowingResult, attachments] =
        await Promise.all([
            getFollowerCount(actorObj),
            getFollowingCount(actorObj),
            isFollowedByDefaultSiteAccount(actorObj, site, accountService),
            getAttachments(actorObj, {
                sanitizeValue: (value: string) => sanitizeHtml(value),
            }),
        ]);

    const accountDto: AccountDTO = {
        id: actor.id,
        name: actor.name || '',
        handle: getHandle(actorObj),
        bio: sanitizeHtml(actor.summary),
        url: actor.url || '',
        avatarUrl: actor.icon?.url || '',
        bannerImageUrl: actor.image?.url || '',
        customFields: {},
        postCount: 0,
        likedCount: 0,
        followingCount: followingCount,
        followerCount: followerCount,
        followedByMe: isFollowingResult,
        followsMe: false,
        attachment: attachments,
    };
    return accountDto;
}
