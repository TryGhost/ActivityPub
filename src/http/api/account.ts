import type { AccountService } from 'account/account.service';
import { getAccountHandle } from 'account/utils';
import type { AppContext } from 'app';
import { sanitizeHtml } from 'helpers/html';
import type { PostService } from 'post/post.service';
import type { AccountDTO } from './types';

/**
 * Maximum number of follow accounts to return
 */
const FOLLOWS_LIMIT = 20;

/**
 * Default number of posts to return in a profile
 */
const DEFAULT_POSTS_LIMIT = 20;

/**
 * Maximum number of posts that can be returned in a profile
 */
const MAX_POSTS_LIMIT = 100;

/**
 * Follow account shape - Used when returning a list of follow accounts
 */
type FollowAccount = Pick<
    AccountDTO,
    'id' | 'name' | 'handle' | 'avatarUrl'
> & { isFollowing: boolean };

/**
 * Create a handler to handle a request for an account
 *
 * @param accountService Account service instance
 */
export function createGetAccountHandler(accountService: AccountService) {
    /**
     * Handle a request for an account
     *
     * @param ctx App context
     */
    return async function handleGetAccount(ctx: AppContext) {
        const logger = ctx.get('logger');
        const site = ctx.get('site');

        // Validate input
        const handle = ctx.req.param('handle') || '';

        if (handle === '') {
            return new Response(null, { status: 400 });
        }

        const db = ctx.get('db');
        let accountDto: AccountDTO;

        const account = await accountService.getDefaultAccountForSite(site);

        if (!account) {
            return new Response(null, { status: 404 });
        }

        try {
            accountDto = {
                /**
                 * At the moment we don't have an internal ID for Ghost accounts so
                 * we use Fediverse ID
                 */
                id: account.ap_id,
                name: account.name || '',
                handle: getAccountHandle(site.host, account.username),
                bio: sanitizeHtml(account.bio || ''),
                url: account.url || '',
                avatarUrl: account.avatar_url || '',
                /**
                 * At the moment we don't support banner images for Ghost accounts
                 */
                bannerImageUrl: account.banner_image_url,
                /**
                 * At the moment we don't support custom fields for Ghost accounts
                 */
                customFields: {},
                postCount: await accountService.getPostCount(account),
                likedCount: await accountService.getLikedCount(account),
                followingCount:
                    await accountService.getFollowingAccountsCount(account),
                followerCount:
                    await accountService.getFollowerAccountsCount(account),
                /**
                 * At the moment we only expect to be returning the account for
                 * the current user, so we can hardcode these values to false as
                 * the account cannot follow, or be followed by itself
                 */
                followsMe: false,
                followedByMe: false,
            };
        } catch (error) {
            logger.error('Error getting account: {error}', { error });

            return new Response(null, { status: 500 });
        }

        // Return response
        return new Response(JSON.stringify(accountDto), {
            headers: {
                'Content-Type': 'application/json',
            },
            status: 200,
        });
    };
}

/**
 * Create a handler to handle a request for a list of account follows
 *
 * @param accountService Account service instance
 */
export function createGetAccountFollowsHandler(accountService: AccountService) {
    /**
     * Handle a request for a list of account follows
     *
     * @param ctx App context
     */
    return async function handleGetAccountFollows(ctx: AppContext) {
        const logger = ctx.get('logger');
        const site = ctx.get('site');

        // Validate input
        const handle = ctx.req.param('handle') || '';

        if (handle === '') {
            return new Response(null, { status: 400 });
        }

        const type = ctx.req.param('type');

        if (!['following', 'followers'].includes(type)) {
            return new Response(null, { status: 400 });
        }

        // Retrieve data
        const getAccounts =
            type === 'following'
                ? accountService.getFollowingAccounts.bind(accountService)
                : accountService.getFollowerAccounts.bind(accountService);
        const getAccountsCount =
            type === 'following'
                ? accountService.getFollowingAccountsCount.bind(accountService)
                : accountService.getFollowerAccountsCount.bind(accountService);

        // @TODO: Get account by provided handle instead of default account?
        const siteDefaultAccount =
            await accountService.getDefaultAccountForSite(site);

        // Get follows accounts and paginate
        const queryNext = ctx.req.query('next') || '0';
        const offset = Number.parseInt(queryNext);

        const results = await getAccounts(siteDefaultAccount, {
            limit: FOLLOWS_LIMIT,
            offset,
            fields: ['id', 'ap_id', 'name', 'username', 'avatar_url'],
        });
        const total = await getAccountsCount(siteDefaultAccount);

        const next =
            total > offset + FOLLOWS_LIMIT
                ? (offset + FOLLOWS_LIMIT).toString()
                : null;

        const accounts: FollowAccount[] = [];

        for (const result of results) {
            accounts.push({
                id: String(result.id),
                name: result.name || '',
                handle: getAccountHandle(
                    new URL(result.ap_id).host,
                    result.username,
                ),
                avatarUrl: result.avatar_url || '',
                isFollowing:
                    type === 'following'
                        ? true
                        : await accountService.checkIfAccountIsFollowing(
                              siteDefaultAccount,
                              result,
                          ),
            });
        }

        // Return response
        return new Response(
            JSON.stringify({
                accounts,
                total,
                next,
            }),
            {
                headers: {
                    'Content-Type': 'application/json',
                },
                status: 200,
            },
        );
    };
}

/**
 * Validates and extracts pagination parameters from the request
 *
 * @param ctx App context
 * @returns Object containing cursor and limit, or null if invalid
 */
function validateRequestParams(ctx: AppContext) {
    const queryCursor = ctx.req.query('next');
    const cursor = queryCursor ? decodeURIComponent(queryCursor) : null;

    const queryLimit = ctx.req.query('limit');
    const limit = queryLimit ? Number(queryLimit) : DEFAULT_POSTS_LIMIT;

    if (limit > MAX_POSTS_LIMIT) {
        return null;
    }

    return { cursor, limit };
}

/**
 * Create a handler to handle a request for a list of posts by an account
 *
 * @param accountService Account service instance
 * @param profileService Profile service instance
 */
export function createGetAccountPostsHandler(
    accountService: AccountService,
    postService: PostService,
) {
    /**
     * Handle a request for a list of posts by an account
     *
     * @param ctx App context
     */
    return async function handleGetPosts(ctx: AppContext) {
        const params = validateRequestParams(ctx);
        if (!params) {
            return new Response(null, { status: 400 });
        }

        const account = await accountService.getDefaultAccountForSite(
            ctx.get('site'),
        );
        const { results, nextCursor } = await postService.getPostsByAccount(
            account.id,
            params.limit,
            params.cursor,
        );

        return new Response(
            JSON.stringify({
                posts: results,
                next: nextCursor,
            }),
            { status: 200 },
        );
    };
}

/**
 * Create a handler to handle a request for a list of posts liked by an account
 *
 * @param accountService Account service instance
 * @param profileService Profile service instance
 */
export function createGetAccountLikedPostsHandler(
    accountService: AccountService,
    postService: PostService,
) {
    /**
     * Handle a request for a list of posts liked by an account
     *
     * @param ctx App context
     */
    return async function handleGetLikedPosts(ctx: AppContext) {
        const params = validateRequestParams(ctx);
        if (!params) {
            return new Response(null, { status: 400 });
        }

        const account = await accountService.getDefaultAccountForSite(
            ctx.get('site'),
        );

        const { results, nextCursor } =
            await postService.getPostsLikedByAccount(
                account.id,
                params.limit,
                params.cursor,
            );

        return new Response(
            JSON.stringify({
                posts: results,
                next: nextCursor,
            }),
            { status: 200 },
        );
    };
}
