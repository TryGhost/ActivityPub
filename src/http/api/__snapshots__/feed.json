{
  "next": null,
  "posts": [
    {
      "attachments": [
        {
          "mediaType": "image/jpeg",
          "name": "test.jpg",
          "type": "Image",
          "url": "https://example.com/test.jpg",
        },
        {
          "mediaType": "image/jpeg",
          "name": "test1.jpg",
          "type": "Image",
          "url": "https://example.com/test1.jpg",
        },
      ],
      "author": {
        "avatarUrl": "",
        "handle": "@foobar@example.com",
        "id": "987",
        "name": "Foo Bar",
        "url": "https://example.com/foobar",
      },
      "content": "Content 789",
      "excerpt": "Excerpt 789",
      "featureImageUrl": "https://example.com/images/post-789.png",
      "id": "https://example.com/.activitypub/post/post-789",
      "likeCount": 1,
      "likedByMe": false,
      "publishedAt": "2025-02-24T16:40:00.000Z",
      "readingTimeMinutes": 3,
      "replyCount": 2,
      "repostCount": 4,
      "repostedBy": null,
      "repostedByMe": false,
      "title": "Post 789",
      "type": 1,
      "url": "https://example.com/post-789",
    },
    {
      "attachments": [],
      "author": {
        "avatarUrl": "",
        "handle": "@foobar@example.com",
        "id": "987",
        "name": "Foo Bar",
        "url": "https://example.com/foobar",
      },
      "content": "Content 790",
      "excerpt": "Excerpt 790",
      "featureImageUrl": "https://example.com/images/post-790.png",
      "id": "https://example.com/.activitypub/post/post-790",
      "likeCount": 0,
      "likedByMe": false,
      "publishedAt": "2025-02-24T16:40:00.000Z",
      "readingTimeMinutes": 2,
      "replyCount": 1,
      "repostCount": 3,
      "repostedBy": {
        "avatarUrl": "https://example.com/images/bazqux.png",
        "handle": "@bazqux@example.com",
        "id": "654",
        "name": "Baz Qux",
        "url": "https://example.com/bazqux",
      },
      "repostedByMe": true,
      "title": "Post 790",
      "type": 1,
      "url": "https://example.com/post-790",
    },
  ],
}